strict-containers/patches/Sequence.patch

```diff
@@ -0,0 +1,46 @@
+Index: strict-containers/src/Data/Strict/Sequence/Autogen/Internal.hs
+===================================================================
+--- strict-containers.orig/src/Data/Strict/Sequence/Autogen/Internal.hs
++++ strict-containers/src/Data/Strict/Sequence/Autogen/Internal.hs
+@@ -995,7 +995,7 @@ seqDataType = mkDataType "Data.Strict.Se
+ 
+ data FingerTree a
+     = EmptyT
+-    | Single a
++    | Single !a
+     | Deep {-# UNPACK #-} !Int !(Digit a) (FingerTree (Node a)) !(Digit a)
+ #ifdef TESTING
+     deriving Show
+@@ -1184,10 +1184,10 @@ pullR s pr m = case viewRTree m of
+ -- Digits
+ 
+ data Digit a
+-    = One a
+-    | Two a a
+-    | Three a a a
+-    | Four a a a a
++    = One !a
++    | Two !a !a
++    | Three !a !a !a
++    | Four !a !a !a !a
```

Comment:
I remember that `Sequence` relies on lazyness to provide (amortized) runtime complexities. I think this change just throws them out of the window - you have to show them new complexities and update the docs.

Reply:
As I understand, the laziness performance gain is on parts of the data structure itself, not of the elements. Specifically, it is via this part, which remains unchanged in this PR:

```haskell
data FingerTree a
  {- ... -}
    | Deep {-# UNPACK #-} !Int !(Digit a) (FingerTree (Node a)) !(Digit a)  -- lazy subtree
```

Reply:
From the paper: "In a strict language that provides a lazy evaluation primitive, we need only suspend the middle subtree of each Deep node, so only Θ(log n) suspensions are required in a tree of size n. Even in a lazy language, some space could be saved in practice by ensuring that these were the only suspensions in the tree, for example by using Haskell’s strictness annotations." i.e. making everything else strict preserves the performance bounds